feature: implement popular movies section in home screen

Noor-Serry opened the pull request:
This commit introduces the popular movies section to the home screen.

Key changes include:


# Pull Request Template

## Description
Provide a clear and concise description of what this pull request does. Include the purpose and context for the changes.
 implement popular movies section in home screen
---

## Changes Made
List the changes introduced in this pull request:

- Added `GetPopularMoviesUseCase` and integrated it into `MovieRepository`.
- Implemented `HomeViewModel` to fetch and manage popular movies state.
- Created `HomeUiState`, `HomeEffect`, and `HomeInteractionListener` for UI logic.
- Developed `HomeUiStateMapper` to transform domain models to UI state.
- Added `PopularMovieCard` composable for displaying individual popular movies.
- Implemented `PopularSection` composable with an auto-scrolling horizontal pager for popular movies.
- Updated `HomeScreen` to display the popular movies section, loading states, and network error handling.
- Added necessary DI modules for the new components.
- Implemented the `getPopularMovies` endpoint in `MovieRemoteSourceImpl`.

---

## Screenshots (if applicable)
Include screenshots or GIFs to showcase the changes, especially if they impact the UI.
https://github.com/user-attachments/assets/28e75608-5bb7-46ac-9e02-5b89782e5f0a

---

## Checklist
Please ensure the following tasks are completed:

- [x] My code follows the code style of this project
- [x] Changes have been tested manually and verified.
- [x] PR includes at most one single feature.

---

## Additional Comments
Add any additional information or context about the pull request here.

Commit:
feature: implement popular movies section in home screen

This commit introduces the popular movies section to the home screen.

Key changes include:
- Added `GetPopularMoviesUseCase` and integrated it into `MovieRepository`.
- Implemented `HomeViewModel` to fetch and manage popular movies state.
- Created `HomeUiState`, `HomeEffect`, and `HomeInteractionListener` for UI logic.
- Developed `HomeUiStateMapper` to transform domain models to UI state.
- Added `PopularMovieCard` composable for displaying individual popular movies.
- Implemented `PopularSection` composable with an auto-scrolling horizontal pager for popular movies.
- Updated `HomeScreen` to display the popular movies section, loading states, and network error handling.
- Added necessary DI modules for the new components.
- Implemented the `getPopularMovies` endpoint in `MovieRemoteSourceImpl`.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/home-popular-section

# Conflicts:
#	remoteDatasource/src/main/java/com/example/remotedatasource/datasource/MovieRemoteSourceImpl.kt
#	repository/src/main/java/com/example/repository/repository/MovieRepositoryImpl.kt

saraashraf11 commented on ui/src/main/java/com/example/ui/screens/home/HomeScreen.kt:
```diff
@@ -1,51 +1,129 @@
+                            onSearchClicked = interactionListener::onClickSearch,
+                            modifier = Modifier
+                                .statusBarsPadding()
+                                .padding(vertical = 0.dp, horizontal = 16.dp),
```
why are you add vertical = 0.dp,?

Belal-Alkhatib commented on viewModel/src/main/java/com/example/viewmodel/home/HomeUiState.kt:
```diff
@@ -0,0 +1,13 @@
+data class HomeUiState(
+    val popularMovies : List<PopularMovie> = emptyList(),
+    val isLoading : Boolean = false,
+    val networkError : Boolean = false
```
we decided to follow a consistent error-handling approach across the app using a sealed interface to represent and map our custom exceptions. So instead of using a simple Boolean like networkError, we should use a sealed error type in HomeUiState.

Commit:
fix: remove vertical padding from HomeTopBar

The vertical padding on the `HomeTopBar` in `HomeScreen.kt` has been removed. The top bar now only has horizontal padding.

Belal-Alkhatib commented on viewModel/src/main/java/com/example/viewmodel/home/HomeUiState.kt:
```diff
@@ -0,0 +1,13 @@
+    val isLoading : Boolean = false,
+    val networkError : Boolean = false
+){
+    data class PopularMovie(
```
for naming consistency across the project, could you please rename this to `PopularMovieItemUiState`? It better reflects its role as a UI state model

eslam-99 commented on ui/src/main/java/com/example/ui/screens/home/component/PopularMovieCard.kt:
```diff
@@ -0,0 +1,92 @@
+
+@ThemeAndLocalePreviews
+@Composable
+fun PopularMovieCardPreview() {
```
Needs private modifier

eslam-99 commented on ui/src/main/java/com/example/ui/screens/home/component/PopularSection.kt:
```diff
@@ -0,0 +1,130 @@
+
+@ThemeAndLocalePreviews
+@Composable
+fun PopularSectionPreview() {
```
Also here

Commit:
chore: make preview functions private

The preview functions for `PopularSection` and `PopularMovieCard` have been made private to improve encapsulation.

eslam-99 reviewed (commented):
Will you add tests for the implemented parts in this PR?
- lets finish the view model first 
- after we finish it will I do the tests but now there are more than one edit in the same view model

ThoriaHamdy99 commented on viewModel/src/main/java/com/example/viewmodel/home/HomeUiStateMapper.kt:
```diff
@@ -0,0 +1,19 @@
+import android.annotation.SuppressLint
+import com.example.entity.Movie
+
+class HomeUiStateMapper {
```
As you made it as a class so, please make an interface for it

Noor-Serry replied:
you ask me to add interface in domain to mapping from domain to ui ?

ThoriaHamdy99 replied:
No interface here for this mapper to be more testable

Noor-Serry replied:
its testable and  we can mock it normally .

ThoriaHamdy99 replied:
But the view model now depends on a concrete class not abstraction if this okay for you its okay but I didn't need the view model to depend on concrete mapper direct.

Noor-Serry replied:
why its not okay. 
not every thing should depend on abstraction

ThoriaHamdy99 replied:
So you need to make it extension if you will leave it

Belal-Alkhatib commented on ui/src/main/java/com/example/ui/screens/home/HomeScreen.kt:
```diff
@@ -1,51 +1,129 @@
+            it?.let {
+                when (it) {
+                    HomeEffect.NavigateToSearchScreenEffect ->
+                        navController.navigate(Route.Search)
```
using navController.navigate(Route.Search) may lead to duplicate destinations in the back stack.
I’ve added a safeNavigate extension function in this PR [#164](https://github.com/Amsterdam-Team/Aflami/pull/164) to handle this case safely. Please consider using it here.

Noor-Serry replied:
its not merged yet

Commit:
refactor: enhance error handling in HomeScreen

This commit refactors the error handling mechanism in `HomeScreen` and `HomeViewModel`.

- `HomeUiState` now includes a sealed class `HomeError` with a `NetworkError` object, replacing the `networkError` boolean.
- `HomeViewModel` updates the state with `HomeError.NetworkError` when a `NoInternetException` occurs.
- `HomeScreen` now checks for `state.error != null` and uses a `when` statement to handle specific error types, currently only `HomeError.NetworkError`.

Commit:
refactor: rename PopularMovie to PopularMovieItemUiState

This commit renames the `PopularMovie` data class within `HomeUiState` to `PopularMovieItemUiState` for better clarity and consistency.

The following files were updated to reflect this change:
- `PopularSection.kt`
- `HomeScreenRoute.kt` (removed unused import)
- `HomeUiStateMapper.kt`
- `HomeUiState.kt`
- `PopularMovieCard.kt`

eslam-99 commented on ui/src/main/java/com/example/ui/screens/home/component/PopularSection.kt:
```diff
@@ -0,0 +1,130 @@
+
+@SuppressLint("RestrictedApi", "ConfigurationScreenWidthHeight", "UnusedBoxWithConstraintsScope")
+fun LazyListScope.PopularSection(popularMovies: List<PopularMovieItemUiState>) {
+    val pagerState = PagerState(currentPage = Int.MAX_VALUE / 2) { Int.MAX_VALUE }
```
Is it better to hoist this state to view model?

Noor-Serry replied:
what state ? pager State !!?

eslam-99 replied:
yes, would we need it later to take action on it from viewmodel?

Noor-Serry replied:
page state its related to android like Lazy State

eslam-99 replied:
Okay

eslam-99 commented on viewModel/src/main/java/com/example/viewmodel/home/HomeInteractionListener.kt:
```diff
@@ -0,0 +1,6 @@
+package com.example.viewmodel.home
+
+interface HomeInteractionListener {
+    fun onClickRetrySearch()
```
No search in home screen, you may mean `onClickRetryLoading` or anything more descriptive

Noor-Serry replied:
I will sovle it

eslam-99 commented on viewModel/src/main/java/com/example/viewmodel/home/HomeViewModel.kt:
```diff
@@ -0,0 +1,55 @@
+    private fun getPopularMovies() {
+        updateState { it.copy(isLoading = true) }
+        tryToExecute(
+            action = { getPopularMoviesUseCase() },
```
Why not passing reference as others?

Noor-Serry replied:
its will not add value this is just one line

eslam-99 replied:
Just consistency, no coding value

Noor-Serry replied:
its okay for me I did not see value from creating other prive function that just call function

Commit:
refactor: rename `onClickRetrySearch` to `onClickRetryLoading`

This commit renames the `onClickRetrySearch` function to `onClickRetryLoading` across the `HomeScreen`, `HomeViewModel`, and `HomeInteractionListener`.

This change improves clarity by accurately reflecting that the retry action is for loading data, not specifically for a search operation.
